cpp/src/io/parquet/page_enc.cu

```diff
@@ -607,34 +672,66 @@ inline __device__ void PackLiterals(
 {
   if (t <= (count | 0x1f)) {
     if (w < 8) {
-      uint32_t mask;
-      if (w <= 1) {
+      uint32_t mask = 0;
+      if (w == 1) {
         v |= SHFL_XOR(v, 1) << 1;
         v |= SHFL_XOR(v, 2) << 2;
         v |= SHFL_XOR(v, 4) << 4;
         mask = 0x7;
-      } else if (w <= 2) {
+      } else if (w == 2) {
         v |= SHFL_XOR(v, 1) << 2;
         v |= SHFL_XOR(v, 2) << 4;
         mask = 0x3;
-      } else {  // w=4
+      } else if (w == 4) {
         v |= SHFL_XOR(v, 1) << 4;
         mask = 0x1;
       }
-      if (t < count && !(t & mask)) { dst[(t * w) >> 3] = v; }
-    } else if (w < 12) {  // w=8
+      if (t < count && mask && !(t & mask)) { dst[(t * w) >> 3] = v; }
+
+      // If w is in set{3,5,6,7}, our job is not done yet.
+      if (w == 1 || w == 2 || w == 4) { return; }
+    } else if (w == 8) {
       if (t < count) { dst[t] = v; }
-    } else if (w < 16) {  // w=12
+      return;
+    } else if (w == 12) {
       v |= SHFL_XOR(v, 1) << 12;
       if (t < count && !(t & 1)) {
         dst[(t >> 1) * 3 + 0] = v;
         dst[(t >> 1) * 3 + 1] = v >> 8;
         dst[(t >> 1) * 3 + 2] = v >> 16;
       }
-    } else if (t < count) {  // w=16
-      dst[t * 2 + 0] = v;
-      dst[t * 2 + 1] = v >> 8;
+      return;
+    } else if (w == 16) {
+      if (t < count) {
+        dst[t * 2 + 0] = v;
+        dst[t * 2 + 1] = v >> 8;
+      }
+      return;
     }
+
+    // Scratch space to temporarily write to. Needed because we will use atomics to write 32 bit
+    // words but the destination mem may not be a multiple of 4 bytes.
+    // TODO (dm): This assumes blockdim = 128 and max bits per value = 16. Reduce magic numbers.
+    __shared__ uint32_t scratch[64];
+    if (t < 64) { scratch[t] = 0; }
+
```

Comment:
Throughout cuio, it is assumed that allocations are aligned and padded to 8 bytes (guaranteed by rmm), so it should be possible to use atomicAnd+atomicOr directly to the output (much less code)

Reply:
I'm not so sure if I'm allowed to make that assumption even if it is correct now. What if something changes somewhere else and this thing breaks? Or someone relocates this method somewhere else in libcudf where because it has wider use.

Reply:
Up to you. IMO, that assumption is safe within libcudf (as long as memory is allocated by rmm). It could be made more explicit by aligning the max allocation size or in writer_impl (generally, it's pretty much a necessary assumption for efficient string comparison functions since GPUs currently lack unaligned 32b accesses, not as important here).